File: src/rename/proximity.ts
/**
 * Proximity windowing for usedNames in large scopes: when a scope has many
 * bindings, prompts only carry the preserved names declared or referenced
 * near the batch's lines, plus well-known globals.
 */
import type { IsEligibleFn } from "./rename-eligibility.js";

/** Looser binding type for proximity windowing (only needs loc info, not path). */
export interface ProximityBinding {
  identifier?: { loc?: { start?: { line?: number } } | null };
  referencePaths?: Array<{
    node?: { loc?: { start?: { line?: number } } | null };
  }>;
}

/** Well-known names that should always appear in usedNames regardless of proximity */
const WELL_KNOWN_NAMES = new Set([
  "exports",
  "require",
  "module",
  "__filename",
  "__dirname",
  "console",
  "process",
  "Buffer",
  "Promise",
  "Object",
  "Array",
  "Map",
  "Set",
  "Error",
  "JSON",
  "Math",
  "undefined",
  "null",
  "NaN",
  "Infinity",
  "setTimeout",
  "setInterval",
  "clearTimeout",
  "clearInterval",
  "parseInt",
  "parseFloat",
  "isNaN",
  "isFinite",
  "encodeURI",
  "decodeURI",
  "encodeURIComponent",
  "decodeURIComponent"
]);

/** Minimum scope bindings before activating proximity windowing */
const WINDOWING_THRESHOLD = 100;

/** Line proximity radius for usedNames windowing */
const PROXIMITY_RADIUS = 100;

/**
 * Returns true if a name is within the proximity window given the binding info.
 */
function isNameInProximityWindow(
  _name: string,
  binding: ProximityBinding | undefined,
  minLine: number,
  maxLine: number,
  alreadyIncluded: boolean
): boolean {
  if (alreadyIncluded) return false;
  if (!binding) return true; // include if binding not found, to be safe

  const declLine = binding.identifier?.loc?.start?.line;
  if (declLine !== undefined && declLine >= minLine && declLine <= maxLine) {
    return true;
  }

  if (binding.referencePaths) {
    for (const refPath of binding.referencePaths) {
      const refLine = refPath.node?.loc?.start?.line;
      if (refLine !== undefined && refLine >= minLine && refLine <= maxLine) {
        return true;
      }
    }
  }

  return false;
}

/**
 * Computes a proximity-windowed subset of usedNames for module-level prompts.
 *
 * When the scope has >= WINDOWING_THRESHOLD bindings, only returns names whose
 * own declarations/references fall within +-PROXIMITY_RADIUS lines of the batch's
 * relevant lines. Always includes well-known names and excludes minified-looking names.
 *
 * When the scope has fewer bindings, returns all non-minified names.
 */
export function getProximateUsedNames(
  allUsedNames: Set<string>,
  batchLines: number[],
  scopeBindings: Record<string, ProximityBinding>,
  totalBindings: number,
  isEligible: IsEligibleFn
): Set<string> {
  const result = new Set<string>();

  // Always include well-known names that are in scope
  for (const name of allUsedNames) {
    if (WELL_KNOWN_NAMES.has(name)) {
      result.add(name);
    }
  }

  // Filter out eligible names (they'll be renamed) — keep only preserved names
  const preserved = [...allUsedNames].filter((n) => !isEligible(n));

  // If below threshold, return all non-minified names
  if (totalBindings < WINDOWING_THRESHOLD) {
    for (const name of preserved) {
      result.add(name);
    }
    return result;
  }

  // Compute the proximity window
  const minLine = Math.min(...batchLines) - PROXIMITY_RADIUS;
  const maxLine = Math.max(...batchLines) + PROXIMITY_RADIUS;

  for (const name of preserved) {
    if (
      isNameInProximityWindow(
        name,
        scopeBindings[name],
        minLine,
        maxLine,
        result.has(name)
      )
    ) {
      result.add(name);
    }
  }

  return result;
}

File: crates/humanify-core/src/rename/votes/proximity.rs
//! Proximity windowing for usedNames in large scopes (WP3.3) — TS
//! original: `src/rename/proximity.ts`.
//!
//! When a scope has many bindings, module-level prompts only carry the
//! PRESERVED (non-eligible) names declared or referenced near the batch's
//! lines, plus the well-known globals. Ported here with WP3.3 (the doc-10
//! placement); its consumer is the wave processor's prompt assembly
//! (WP4.3), which supplies the per-name line data.

/// A name's binding as the window reads it: the declaration line and the
/// reference lines (TS `ProximityBinding` — both optional there, `None` /
/// empty here).
#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct ProximityBinding {
    pub decl_line: Option<u32>,
    pub ref_lines: Vec<u32>,
}

/// Names always kept regardless of proximity.
pub const WELL_KNOWN_NAMES: [&str; 32] = [
    "exports",
    "require",
    "module",
    "__filename",
    "__dirname",
    "console",
    "process",
    "Buffer",
    "Promise",
    "Object",
    "Array",
    "Map",
    "Set",
    "Error",
    "JSON",
    "Math",
    "undefined",
    "null",
    "NaN",
    "Infinity",
    "setTimeout",
    "setInterval",
    "clearTimeout",
    "clearInterval",
    "parseInt",
    "parseFloat",
    "isNaN",
    "isFinite",
    "encodeURI",
    "decodeURI",
    "encodeURIComponent",
    "decodeURIComponent",
];

/// Minimum scope bindings before windowing activates.
pub const WINDOWING_THRESHOLD: usize = 100;
/// Line proximity radius.
pub const PROXIMITY_RADIUS: f64 = 100.0;

/// TS `isNameInProximityWindow`.
fn in_window(binding: Option<&ProximityBinding>, min_line: f64, max_line: f64) -> bool {
    let Some(binding) = binding else {
        // Include when the binding is not found, to be safe.
        return true;
    };
    let within = |line: u32| {
        let l = f64::from(line);
        l >= min_line && l <= max_line
    };
    binding.decl_line.is_some_and(within) || binding.ref_lines.iter().any(|&l| within(l))
}

/// TS `getProximateUsedNames`: the windowed usedNames, in insertion order
/// (the TS returns a Set — well-known names first, then the preserved
/// names in `all_used_names` order).
pub fn get_proximate_used_names<'a, S: AsRef<str>>(
    all_used_names: &'a [S],
    batch_lines: &[u32],
    scope_binding: impl Fn(&str) -> Option<ProximityBinding>,
    total_bindings: usize,
    is_eligible: impl Fn(&str) -> bool,
) -> Vec<String> {
    // The Set: insertion-ordered result + a membership index (the waves
    // call this per request over ~25k names — a linear `has` is quadratic).
    let mut result: Vec<String> = Vec::new();
    let mut members: std::collections::HashSet<&str> = std::collections::HashSet::new();
    let names = all_used_names.iter().map(AsRef::as_ref);
    let mut push = |result: &mut Vec<String>, name: &'a str| {
        if members.insert(name) {
            result.push(name.to_string());
        }
    };
    for name in names.clone() {
        if WELL_KNOWN_NAMES.contains(&name) {
            push(&mut result, name);
        }
    }
    let preserved: Vec<&str> = names.filter(|n| !is_eligible(n)).collect();
    if total_bindings < WINDOWING_THRESHOLD {
        for name in preserved {
            push(&mut result, name);
        }
        return result;
    }
    // Math.min(...[]) is Infinity and Math.max(...[]) -Infinity: an empty
    // batch windows nothing in.
    let min_line = batch_lines
        .iter()
        .map(|&l| f64::from(l))
        .fold(f64::INFINITY, f64::min)
        - PROXIMITY_RADIUS;
    let max_line = batch_lines
        .iter()
        .map(|&l| f64::from(l))
        .fold(f64::NEG_INFINITY, f64::max)
        + PROXIMITY_RADIUS;
    for name in preserved {
        // (a member already in the result is skipped — `alreadyIncluded`)
        // `ownEntry(scopeBindings, name)`: an absent name — including one
        // named after an Object.prototype member — is absent, so it is
        // included "to be safe" (16-findings-queue #22, fixed TS-first).
        let binding = scope_binding(name);
        if in_window(binding.as_ref(), min_line, max_line) {
            push(&mut result, name);
        }
    }
    result
}

#[cfg(test)]
mod proximity_test;
